implement cobra for a better cli

FireFart opened the pull request:
This implements [cobra](https://github.com/spf13/cobra) for more convenient cli options and to make implementations of future plugins more easy.

Changes:
* each plugin now has specific cli options. no more mixing of options
* there are now command instead of the mode switch. Now you have to do `gobuster dir` or `gobuster dns`
* switches with 2 characters have been renamed or have no more short form as cobra does not support it
* the interface methods have been renamed to smth more meaningful (`PreRun` and `Run`)
* as some parameter names have changed and there are now commands this change is not backward compatible. So scripts calling gobuster need to be adopted

PS: most of the code was copied into the plugins (with slight modifications as some options now reside in the plugin object), so it's not that much changes as github shows :)

OJ commented:
Goddammit these huge PRs are a PITA to review :P

OJ commented on cli/cmd/root.go:
```diff
@@ -0,0 +1,84 @@
+	rootCmd.PersistentFlags().StringP("output", "o", "", "Output file to write results to (defaults to stdout)")
+	rootCmd.PersistentFlags().BoolP("verbose", "v", false, "Verbose output (errors)")
+	rootCmd.PersistentFlags().BoolP("quiet", "q", false, "Don't print the banner and other noise")
+	rootCmd.PersistentFlags().BoolP("noprogress", "", false, "Don't display progress")
```
Previously we had `-np` for this, does this mean that we can't use that any more? I assume it has to be `--noprogress` ?

FireFart replied:
jeah cobra only allows one letter short options :( but we can use another letter if you like like `-n`

FireFart replied:
oh `-n` is already taken so it has to be another one. but we can change other options too if you like, now it's the chance :)

OJ commented on gobusterdir/gobusterdir.go:
```diff
@@ -129,3 +167,99 @@ func (d GobusterDir) ResultToString(g *libgobuster.Gobuster, r *libgobuster.Resu
+	if d.globalopts.Wordlist != "-" {
+		wordlist = d.globalopts.Wordlist
+	}
+	if _, err := fmt.Fprintf(buf, "[+] Wordlist     : %s\n", wordlist); err != nil {
```
Picky, but missing newline before this line :)

OJ commented on gobusterdns/gobusterdns.go:
```diff
@@ -93,3 +117,57 @@ func (d GobusterDNS) ResultToString(g *libgobuster.Gobuster, r *libgobuster.Resu
+	if _, err := fmt.Fprintf(buf, "[+] Domain         : %s\n", o.Domain); err != nil {
+		return "", err
+	}
+	if _, err := fmt.Fprintf(buf, "[+] Threads        : %d\n", d.globalopts.Threads); err != nil {
```
Another missing newline nitpick. Sorry :D

OJ commented:
We'll have to update the `README` as well so that all the new flags match (eg. the `--noprogress` option is different).

FireFart commented:
I updated the readme but you have to change the examples once we release a new version (because of your wordlist). I updated all commands so you just need to copy paste them to your terminal

OJ commented:
OK going to take a look at this in a sec. Given that we're changing the CLI in this PR, I think we're going to have to bump the major version number. In doing so, we may as well get a few other things merged in at the same time. So we'll continue to merge into the current working branch, but we'll release as 3.0 when ready.

OJ commented:
I've dropped this into the new `v3.0-working` branch at the moment. There are a few other things I want to merge in, and there's still an issue I want to resolve as well (particularly, when you press CTRL+C, I want the `context cancelled` error messages to not appear (URL mode).

OJ commented:
So you know, `v3.0-working` was branched off `v2.0.2-working` so we haven't lost anything.

I'm going to close this PR now because it's technically landed in that branch. More work to do though. Thanks!

FireFart commented:
@OJ thanks! Yeah doing a 3.0.0 was the plan (see here https://github.com/OJ/gobuster/pull/117). So would you like me to push everything we have so far to the new v3 branch and test everything together or should we do one PR after another into the new v3 branch?

FireFart commented:
ok thanks! The order of the remaining PRs is:
- https://github.com/OJ/gobuster/pull/101
- https://github.com/OJ/gobuster/pull/104
- https://github.com/OJ/gobuster/pull/107
- https://github.com/OJ/gobuster/pull/117

The diffs now look way cleaner after this beast is merged :)
